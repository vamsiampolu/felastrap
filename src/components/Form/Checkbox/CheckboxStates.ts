export type CheckboxStates = "Checked" | "Indeterminate" | "Empty";
export const CHECKBOX_STATES: Record<CheckboxStates, CheckboxStates> = {
  Checked: "Checked",
  Indeterminate: "Indeterminate",
  Empty: "Empty",
};
