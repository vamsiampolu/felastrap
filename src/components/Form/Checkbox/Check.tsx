import { ComponentPropsWithoutRef } from "react";
import { useFela } from "react-fela";
import { StyleProps } from "../../common";
import { CHECKBOX_STATES, CheckboxStates } from "./CheckboxStates";
import { useCheckbox } from "./useCheckbox";
import checked from "/checked.svg";
import indeterminate from "/indeterminate.svg";

type RuleProps = StyleProps<{ reverse: boolean }>;

const rule = () => ({
  float: "left",
  marginLeft: "-1.5em",
  flexShrink: 0,
  width: "1em",
  height: "1em",
  marginTop: "0.25em",
  verticalAlign: "top",
  appearance: "none",
  backgroundColor: "#fff",
  backgroundRepeat: "no-repeat",
  backgroundPosition: "center",
  backgroundSize: "contain",
  border: "1px solid #dee2e6",
  colorAdjust: "exact",
  printColorAdjust: "exact",
  borderRadius: "0.25em",
  ":active": {
    filter: "brightness(90%)",
  },
  ":checked": {
    backgroundColor: "#0d6efd",
    backgroundImage: `url(${checked})`,
    borderColor: " #0d6efd",
  },
  ":focus": {
    borderColor: "#86b7fe",
    outline: 0,
    boxShadow: "0 0 0 0.25rem rgba(13, 110, 253, 0.25)",
  },
  ":indeterminate": {
    backgroundColor: "#0d6efd",
    borderColor: "#0d6efd",
    backgroundImage: `url(${indeterminate})`,
  },
  ":disabled": {
    pointerEvents: "none",
    filter: "none",
    opacity: 0.5,
  },
});

const reverseRule = ({ reverse }: RuleProps) =>
  reverse
    ? {
        float: "right",
        marginRight: "-1.5em",
        marginLeft: 0,
      }
    : {
        float: "left",
        marginRight: 0,
        marginLeft: "-1.5em",
      };

export type Attributes =
  | "id"
  | "name"
  | "className"
  | "checked"
  | "disabled"
  | "defaultValue"
  | "onSelect"
  | "onChange"
  | "onInput"
  | "required";

export type CheckboxInputProps = Pick<
  ComponentPropsWithoutRef<"input">,
  Attributes
> & {
  reverse?: boolean;
  value?: CheckboxStates;
  defaultValue?: CheckboxStates;
  allowIndeterminate?: boolean;
};

export const CheckboxInput = ({
  id,
  className = "",
  disabled = false,
  reverse = false,
  value = CHECKBOX_STATES.Empty,
  allowIndeterminate = false,
  onChange = () => undefined,
  ...props
}: CheckboxInputProps) => {
  const ref = useCheckbox({ value, allowIndeterminate });
  const { css } = useFela({ reverse });
  const cx = `${css([rule, reverseRule])} form-check-input ${className}`.trim();

  return (
    <input
      data-testid="checkbox-input"
      type="checkbox"
      disabled={disabled}
      id={id}
      className={cx}
      ref={ref}
      value={value}
      onChange={onChange}
      {...props}
    />
  );
};
