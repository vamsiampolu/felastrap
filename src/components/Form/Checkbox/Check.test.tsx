import { IRenderer } from "fela";
import { CheckboxInput } from "./Check";
import {
  Wrapper,
  createRenderer,
  createSnapshot,
  cssReboot,
} from "../../../test-helpers";
import { render, screen } from "@testing-library/react";
import { CHECKBOX_STATES } from "./CheckboxStates";
import userEvent from "@testing-library/user-event";

describe("components/Form/Checkbox/Check", () => {
  let renderer: IRenderer;

  beforeEach(() => {
    renderer = createRenderer();
    cssReboot(renderer);
  });

  afterEach(() => {
    renderer.clear();
  });

  it("should correctly render a default checkbox input", async () => {
    const component = <CheckboxInput />;
    render(<Wrapper renderer={renderer}>{component}</Wrapper>);

    expect(screen.getByTestId("checkbox-input")).toBeInTheDocument();
    expect(screen.queryByTestId("checkbox-input")?.tagName).toEqual("INPUT");

    expect(screen.queryByTestId("checkbox-input")).toHaveAttribute(
      "type",
      "checkbox",
    );
    expect(screen.queryByTestId("checkbox-input")).not.toBeChecked();

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should correctly render a disabled checkbox input", async () => {
    const component = <CheckboxInput disabled />;
    render(<Wrapper renderer={renderer}>{component}</Wrapper>);

    expect(screen.getByTestId("checkbox-input")).toBeInTheDocument();
    expect(screen.queryByTestId("checkbox-input")?.tagName).toEqual("INPUT");

    expect(screen.queryByTestId("checkbox-input")).toHaveAttribute(
      "type",
      "checkbox",
    );
    expect(screen.queryByTestId("checkbox-input")).not.toBeChecked();
    expect(screen.queryByTestId("checkbox-input")).toBeDisabled();

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should render a checked checkbox input", async () => {
    const component = (
      <CheckboxInput id="form-check" value={CHECKBOX_STATES.Checked} />
    );
    render(<Wrapper renderer={renderer}>{component}</Wrapper>);

    expect(screen.getByTestId("checkbox-input")).toBeInTheDocument();
    expect(screen.queryByTestId("checkbox-input")).toBeChecked();
    expect(screen.queryByTestId("checkbox-input")).toHaveAttribute(
      "id",
      "form-check",
    );

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should render an indeterminate checkbox input", async () => {
    const component = (
      <CheckboxInput
        id="form-check"
        value={CHECKBOX_STATES.Indeterminate}
        allowIndeterminate
      />
    );
    render(<Wrapper renderer={renderer}>{component}</Wrapper>);

    expect(screen.getByTestId("checkbox-input")).toBeInTheDocument();
    expect(
      (screen.queryByTestId("checkbox-input") as HTMLInputElement)
        .indeterminate,
    ).toBe(true);

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should invoke onChange when checkbox state changes", async () => {
    const onChange = jest.fn();
    const component = <CheckboxInput onChange={onChange} />;

    render(<Wrapper renderer={renderer}>{component}</Wrapper>);
    expect(screen.getByTestId("checkbox-input")).toBeInTheDocument();

    const user = userEvent.setup();
    const checkbox = screen.queryByTestId("checkbox-input");

    await user.click(checkbox!);

    expect(screen.queryByTestId("checkbox-input")).toBeChecked();
    expect(onChange).toHaveBeenCalled();
  });
});
