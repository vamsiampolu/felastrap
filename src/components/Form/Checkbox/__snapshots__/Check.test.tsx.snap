// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Form/Checkbox/Check should correctly render a default checkbox input 1`] = `
".a {
  float: left;
}
.b {
  margin-left: -1.5em;
}
.c {
  flex-shrink: 0;
}
.d {
  width: 1em;
}
.e {
  height: 1em;
}
.f {
  margin-top: 0.25em;
}
.g {
  vertical-align: top;
}
.h {
  appearance: none;
}
.i {
  background-color: #fff;
}
.j {
  background-repeat: no-repeat;
}
.k {
  background-position: center;
}
.l {
  background-size: contain;
}
.m {
  color-adjust: exact;
}
.n {
  print-color-adjust: exact;
}
.p:checked {
  background-color: #0d6efd;
}
.q:checked {
  background-image: url(test-file-stub);
}
.r:checked {
  border-top-color: #0d6efd;
}
.s:checked {
  border-right-color: #0d6efd;
}
.t:checked {
  border-bottom-color: #0d6efd;
}
.u:checked {
  border-left-color: #0d6efd;
}
.ac:indeterminate {
  background-color: #0d6efd;
}
.ae:indeterminate {
  background-image: url(test-file-stub);
}
.af:indeterminate {
  border-top-color: #0d6efd;
}
.ag:indeterminate {
  border-right-color: #0d6efd;
}
.ah:indeterminate {
  border-bottom-color: #0d6efd;
}
.ai:indeterminate {
  border-left-color: #0d6efd;
}
.aj:disabled {
  pointer-events: none;
}
.ak:disabled {
  filter: none;
}
.al:disabled {
  opacity: 0.5;
}
.am {
  margin-right: 0;
}
.an {
  border-top-left-radius: 0.25em;
}
.ao {
  border-top-right-radius: 0.25em;
}
.ap {
  border-bottom-right-radius: 0.25em;
}
.aq {
  border-bottom-left-radius: 0.25em;
}
.ar {
  border-top-width: 1px;
}
.as {
  border-right-width: 1px;
}
.at {
  border-bottom-width: 1px;
}
.au {
  border-left-width: 1px;
}
.av {
  border-top-style: solid;
}
.aw {
  border-right-style: solid;
}
.ax {
  border-bottom-style: solid;
}
.ay {
  border-left-style: solid;
}
.az {
  border-top-color: #dee2e6;
}
.ba {
  border-right-color: #dee2e6;
}
.bb {
  border-bottom-color: #dee2e6;
}
.bc {
  border-left-color: #dee2e6;
}
.v:focus {
  box-shadow: 0 0 0 0.25rem rgba(13, 110, 253, 0.25);
}
.w:focus {
  outline-width: 0;
}
.x:focus {
  border-top-color: #86b7fe;
}
.y:focus {
  border-right-color: #86b7fe;
}
.z:focus {
  border-bottom-color: #86b7fe;
}
.ab:focus {
  border-left-color: #86b7fe;
}
.o:active {
  filter: brightness(90%);
}


<input
  data-testid="checkbox-input"
  type="checkbox"
  class="a ab ac ae af ag ah ai aj ak al am an ao ap aq ar as at au av aw ax ay az b ba bb bc c d e f g h i j k l m n o p q r s t u v w x y z form-check-input"
  value="Empty"
/>;
"
`;

exports[`components/Form/Checkbox/Check should correctly render a disabled checkbox input 1`] = `
".a {
  float: left;
}
.b {
  margin-left: -1.5em;
}
.c {
  flex-shrink: 0;
}
.d {
  width: 1em;
}
.e {
  height: 1em;
}
.f {
  margin-top: 0.25em;
}
.g {
  vertical-align: top;
}
.h {
  appearance: none;
}
.i {
  background-color: #fff;
}
.j {
  background-repeat: no-repeat;
}
.k {
  background-position: center;
}
.l {
  background-size: contain;
}
.m {
  color-adjust: exact;
}
.n {
  print-color-adjust: exact;
}
.p:checked {
  background-color: #0d6efd;
}
.q:checked {
  background-image: url(test-file-stub);
}
.r:checked {
  border-top-color: #0d6efd;
}
.s:checked {
  border-right-color: #0d6efd;
}
.t:checked {
  border-bottom-color: #0d6efd;
}
.u:checked {
  border-left-color: #0d6efd;
}
.ac:indeterminate {
  background-color: #0d6efd;
}
.ae:indeterminate {
  background-image: url(test-file-stub);
}
.af:indeterminate {
  border-top-color: #0d6efd;
}
.ag:indeterminate {
  border-right-color: #0d6efd;
}
.ah:indeterminate {
  border-bottom-color: #0d6efd;
}
.ai:indeterminate {
  border-left-color: #0d6efd;
}
.aj:disabled {
  pointer-events: none;
}
.ak:disabled {
  filter: none;
}
.al:disabled {
  opacity: 0.5;
}
.am {
  margin-right: 0;
}
.an {
  border-top-left-radius: 0.25em;
}
.ao {
  border-top-right-radius: 0.25em;
}
.ap {
  border-bottom-right-radius: 0.25em;
}
.aq {
  border-bottom-left-radius: 0.25em;
}
.ar {
  border-top-width: 1px;
}
.as {
  border-right-width: 1px;
}
.at {
  border-bottom-width: 1px;
}
.au {
  border-left-width: 1px;
}
.av {
  border-top-style: solid;
}
.aw {
  border-right-style: solid;
}
.ax {
  border-bottom-style: solid;
}
.ay {
  border-left-style: solid;
}
.az {
  border-top-color: #dee2e6;
}
.ba {
  border-right-color: #dee2e6;
}
.bb {
  border-bottom-color: #dee2e6;
}
.bc {
  border-left-color: #dee2e6;
}
.v:focus {
  box-shadow: 0 0 0 0.25rem rgba(13, 110, 253, 0.25);
}
.w:focus {
  outline-width: 0;
}
.x:focus {
  border-top-color: #86b7fe;
}
.y:focus {
  border-right-color: #86b7fe;
}
.z:focus {
  border-bottom-color: #86b7fe;
}
.ab:focus {
  border-left-color: #86b7fe;
}
.o:active {
  filter: brightness(90%);
}


<input
  data-testid="checkbox-input"
  type="checkbox"
  disabled=""
  class="a ab ac ae af ag ah ai aj ak al am an ao ap aq ar as at au av aw ax ay az b ba bb bc c d e f g h i j k l m n o p q r s t u v w x y z form-check-input"
  value="Empty"
/>;
"
`;

exports[`components/Form/Checkbox/Check should render a checked checkbox input 1`] = `
".a {
  float: left;
}
.b {
  margin-left: -1.5em;
}
.c {
  flex-shrink: 0;
}
.d {
  width: 1em;
}
.e {
  height: 1em;
}
.f {
  margin-top: 0.25em;
}
.g {
  vertical-align: top;
}
.h {
  appearance: none;
}
.i {
  background-color: #fff;
}
.j {
  background-repeat: no-repeat;
}
.k {
  background-position: center;
}
.l {
  background-size: contain;
}
.m {
  color-adjust: exact;
}
.n {
  print-color-adjust: exact;
}
.p:checked {
  background-color: #0d6efd;
}
.q:checked {
  background-image: url(test-file-stub);
}
.r:checked {
  border-top-color: #0d6efd;
}
.s:checked {
  border-right-color: #0d6efd;
}
.t:checked {
  border-bottom-color: #0d6efd;
}
.u:checked {
  border-left-color: #0d6efd;
}
.ac:indeterminate {
  background-color: #0d6efd;
}
.ae:indeterminate {
  background-image: url(test-file-stub);
}
.af:indeterminate {
  border-top-color: #0d6efd;
}
.ag:indeterminate {
  border-right-color: #0d6efd;
}
.ah:indeterminate {
  border-bottom-color: #0d6efd;
}
.ai:indeterminate {
  border-left-color: #0d6efd;
}
.aj:disabled {
  pointer-events: none;
}
.ak:disabled {
  filter: none;
}
.al:disabled {
  opacity: 0.5;
}
.am {
  margin-right: 0;
}
.an {
  border-top-left-radius: 0.25em;
}
.ao {
  border-top-right-radius: 0.25em;
}
.ap {
  border-bottom-right-radius: 0.25em;
}
.aq {
  border-bottom-left-radius: 0.25em;
}
.ar {
  border-top-width: 1px;
}
.as {
  border-right-width: 1px;
}
.at {
  border-bottom-width: 1px;
}
.au {
  border-left-width: 1px;
}
.av {
  border-top-style: solid;
}
.aw {
  border-right-style: solid;
}
.ax {
  border-bottom-style: solid;
}
.ay {
  border-left-style: solid;
}
.az {
  border-top-color: #dee2e6;
}
.ba {
  border-right-color: #dee2e6;
}
.bb {
  border-bottom-color: #dee2e6;
}
.bc {
  border-left-color: #dee2e6;
}
.v:focus {
  box-shadow: 0 0 0 0.25rem rgba(13, 110, 253, 0.25);
}
.w:focus {
  outline-width: 0;
}
.x:focus {
  border-top-color: #86b7fe;
}
.y:focus {
  border-right-color: #86b7fe;
}
.z:focus {
  border-bottom-color: #86b7fe;
}
.ab:focus {
  border-left-color: #86b7fe;
}
.o:active {
  filter: brightness(90%);
}


<input
  data-testid="checkbox-input"
  type="checkbox"
  id="form-check"
  class="a ab ac ae af ag ah ai aj ak al am an ao ap aq ar as at au av aw ax ay az b ba bb bc c d e f g h i j k l m n o p q r s t u v w x y z form-check-input"
  value="Checked"
/>;
"
`;

exports[`components/Form/Checkbox/Check should render an indeterminate checkbox input 1`] = `
".a {
  float: left;
}
.b {
  margin-left: -1.5em;
}
.c {
  flex-shrink: 0;
}
.d {
  width: 1em;
}
.e {
  height: 1em;
}
.f {
  margin-top: 0.25em;
}
.g {
  vertical-align: top;
}
.h {
  appearance: none;
}
.i {
  background-color: #fff;
}
.j {
  background-repeat: no-repeat;
}
.k {
  background-position: center;
}
.l {
  background-size: contain;
}
.m {
  color-adjust: exact;
}
.n {
  print-color-adjust: exact;
}
.p:checked {
  background-color: #0d6efd;
}
.q:checked {
  background-image: url(test-file-stub);
}
.r:checked {
  border-top-color: #0d6efd;
}
.s:checked {
  border-right-color: #0d6efd;
}
.t:checked {
  border-bottom-color: #0d6efd;
}
.u:checked {
  border-left-color: #0d6efd;
}
.ac:indeterminate {
  background-color: #0d6efd;
}
.ae:indeterminate {
  background-image: url(test-file-stub);
}
.af:indeterminate {
  border-top-color: #0d6efd;
}
.ag:indeterminate {
  border-right-color: #0d6efd;
}
.ah:indeterminate {
  border-bottom-color: #0d6efd;
}
.ai:indeterminate {
  border-left-color: #0d6efd;
}
.aj:disabled {
  pointer-events: none;
}
.ak:disabled {
  filter: none;
}
.al:disabled {
  opacity: 0.5;
}
.am {
  margin-right: 0;
}
.an {
  border-top-left-radius: 0.25em;
}
.ao {
  border-top-right-radius: 0.25em;
}
.ap {
  border-bottom-right-radius: 0.25em;
}
.aq {
  border-bottom-left-radius: 0.25em;
}
.ar {
  border-top-width: 1px;
}
.as {
  border-right-width: 1px;
}
.at {
  border-bottom-width: 1px;
}
.au {
  border-left-width: 1px;
}
.av {
  border-top-style: solid;
}
.aw {
  border-right-style: solid;
}
.ax {
  border-bottom-style: solid;
}
.ay {
  border-left-style: solid;
}
.az {
  border-top-color: #dee2e6;
}
.ba {
  border-right-color: #dee2e6;
}
.bb {
  border-bottom-color: #dee2e6;
}
.bc {
  border-left-color: #dee2e6;
}
.v:focus {
  box-shadow: 0 0 0 0.25rem rgba(13, 110, 253, 0.25);
}
.w:focus {
  outline-width: 0;
}
.x:focus {
  border-top-color: #86b7fe;
}
.y:focus {
  border-right-color: #86b7fe;
}
.z:focus {
  border-bottom-color: #86b7fe;
}
.ab:focus {
  border-left-color: #86b7fe;
}
.o:active {
  filter: brightness(90%);
}


<input
  data-testid="checkbox-input"
  type="checkbox"
  id="form-check"
  class="a ab ac ae af ag ah ai aj ak al am an ao ap aq ar as at au av aw ax ay az b ba bb bc c d e f g h i j k l m n o p q r s t u v w x y z form-check-input"
  value="Indeterminate"
/>;
"
`;
